### PR TITLE
Add artifactory support

### PR DESCRIPTION
**Pull request by jdekonin**
* adds the capability of reducing footprint on jenkins jobs
    concerning the storage of artifacts

Signed-off-by: Joe deKoning <joe_dekoning@ca.ibm.com>

### CODE REVIEW

**Comment by jdekonin** [comment]
fyi @AdamBrousseau

**Comment by AdamBrousseau** [comment]
May need to figure out tagging levels shortly after we merge this. Anyone wanting to pull "nightlies" will need to pull the right level and not a developer build.

**Comment by DanHeidinga** [comment]
> Anyone wanting to pull "nightlies" will need to pull the right level and not a developer build.

Looking for some clarification - this is in regards to the build system pulling specific builds to run tests on, correct?

OpenJ9 is a source only project and doesn't provide releases directly,  Binary builds are available from the AdoptOpenJDK project.

**Comment by AdamBrousseau** [comment]
>this is in regards to the build system pulling specific builds

No, the build system is fine and will know what to pull. My comment is regarding anyone, manually or automatically, wanting to pull binaries outside of our build system for other testing purposes. They currently won't be able to determine what "build" they're getting.

File: buildenv/jenkins/common/build [hunk 1]
```diff
@@ -232,10 +232,36 @@ def build() {
...
-            archiveArtifacts artifacts: "**/${TEST_PREFIX}*${TEST_SUFFIX}", fingerprint: true, onlyIfSuccessful: true
-            sh "tar -C build/$RELEASE/images/ -zcvf ${SDK_PREFIX}`date +%Y%d%m%H%M`${SDK_SUFFIX} ${JDK_FOLDER}"
-            archiveArtifacts artifacts: "**/${SDK_PREFIX}*${SDK_SUFFIX}", fingerprint: true, onlyIfSuccessful: true
+              SDK_FILENAME = "${SDK_PREFIX}${env.BUILD_ID}${SDK_SUFFIX}"
+              TESTSHA = sh (
+                  script: 'git -C openj9 rev-parse --short HEAD',
+                  returnStdout: true
+              ).trim()
+              TEST_FILENAME = "${TEST_PREFIX}${TESTSHA}${TEST_SUFFIX}"
+              sh "tar -zcvf ${TEST_FILENAME} openj9/test/"
```
[the remark this answers is not in the document]
Comment (jdekonin) [reply buildenv/jenkins/common/build]: I was going to clarify that with @llxia that the test.jar isn't required anymore.  I believe the only jobs that depend on that any longer are JDK9 which @pshipton has asked to be turned off from nightly builds.

Update: There are also a few windows jobs, but those need to be updated to work in the openjdk-test manner.

File: buildenv/jenkins/common/build [hunk 2]
```diff
@@ -232,10 +232,36 @@ def build() {
 def archive() {
     stage('Archive') {
         timestamps {
-            sh "tar -zcvf ${TEST_PREFIX}`git -C openj9 rev-parse --short HEAD`${TEST_SUFFIX} openj9/test/"
-            archiveArtifacts artifacts: "**/${TEST_PREFIX}*${TEST_SUFFIX}", fingerprint: true, onlyIfSuccessful: true
-            sh "tar -C build/$RELEASE/images/ -zcvf ${SDK_PREFIX}`date +%Y%d%m%H%M`${SDK_SUFFIX} ${JDK_FOLDER}"
-            archiveArtifacts artifacts: "**/${SDK_PREFIX}*${SDK_SUFFIX}", fingerprint: true, onlyIfSuccessful: true
+              SDK_FILENAME = "${SDK_PREFIX}${env.BUILD_ID}${SDK_SUFFIX}"
```
[the remark this answers is not in the document]
Comment (jdekonin) [reply buildenv/jenkins/common/build]: I can revert, I did that to keep the filename simple.  The upload and directory naming need a conversation at a broader level once put into use.  Having a datestamp in the filename attached as a jenkins artifact is irrelevant as all that information is attached to the job.

File: buildenv/jenkins/common/build [hunk 3]
```diff
@@ -232,10 +232,36 @@ def build() {
...
+                    {
+                      "pattern": "${env.WORKSPACE}/*gz",
+                      "target": "${ARTIFACTORY_REPO}/${env.JOB_NAME}/${env.BUILD_ID}/"
+                    }
+                  ]
+                }"""
+
+                def buildInfo = Artifactory.newBuildInfo()
+                // hardcoded to 20 for now, this need fup for usage on per platform level or across all jobs
+                buildInfo.retention maxBuilds: 30, deleteBuildArtifacts: true
```
[the remark this answers is not in the document]
Comment (jdekonin) [reply buildenv/jenkins/common/build]: Yes.

**Comment by jdekonin** [comment]
@AdamBrousseau I believe this is now ready for review.  I can walk you through the testing that I performed in sandbox jobs if you want to see the changes in action.

File: buildenv/jenkins/common/variables-functions [hunk 4]
```diff
@@ -378,6 +368,26 @@ def set_build_variables() {
...
+    if (!SDK_FILENAME) {
+        try{
+            SDK_PREFIX = "OpenJ9-JDK${SDK_VERSION}-PR${ghprbPullId}-${SPEC}-"
+        } catch (e) {
+            SDK_PREFIX = "OpenJ9-JDK${SDK_VERSION}-${SPEC}-"
+        }
+
+        SDK_SUFFIX = ".tar.gz"
+        TEST_PREFIX = "test-"
+        TEST_SUFFIX = ".tar.gz"
```
Comment (AdamBrousseau) [inline buildenv/jenkins/common/variables-functions]: These are no longer needed

File: buildenv/jenkins/common/variables-functions [hunk 5]
```diff
@@ -378,6 +368,26 @@ def set_build_variables() {
     echo "Using BUILD_ENV_CMD = ${BUILD_ENV_CMD}, BUILD_ENV_VARS_LIST = ${BUILD_ENV_VARS_LIST.toString()}"
 }
 
+def set_sdk_variables() {
+    SDK_FILENAME = params.SDK_FILENAME
+    if (!SDK_FILENAME) {
+        try{
+            SDK_PREFIX = "OpenJ9-JDK${SDK_VERSION}-PR${ghprbPullId}-${SPEC}-"
```
Comment (AdamBrousseau) [inline buildenv/jenkins/common/variables-functions]: We don't archive PRs anymore so we might as well remove this while we're here. Can likely merge Prefix and Suffix into the SDK_FILENAME declaration

Comment (jdekonin) [reply buildenv/jenkins/common/variables-functions]: Changes made in latest push

File: buildenv/jenkins/common/build [hunk 6]
```diff
@@ -261,10 +261,28 @@ def build() {
...
+                    {
+                      "pattern": "${env.WORKSPACE}/${SDK_FILENAME}",
+                      "target": "${ARTIFACTORY_REPO}/${env.JOB_NAME}/${env.BUILD_ID}/"
+                    }
+                  ]
+                }"""
+
+                def buildInfo = Artifactory.newBuildInfo()
+                // hardcoded to 30 for now, this need follow up for usage on per platform level or across all jobs
+                buildInfo.retention maxBuilds: 30, deleteBuildArtifacts: true
```
Comment (AdamBrousseau) [inline buildenv/jenkins/common/build]: Can probably use this once its ready (just fyi, not blocking this on that)
https://github.com/eclipse/openj9/pull/2138/files#diff-86d1cc0920536c6d4819089e4772aa71R472
https://github.com/eclipse/openj9/pull/2138/files#diff-a3fca431c8c372cebd1caff040ca6eabR59

File: buildenv/jenkins/common/variables-functions [hunk 7]
```diff
@@ -332,17 +332,7 @@ def set_build_variables() {
...
-        SDK_PREFIX = "OpenJ9-JDK${SDK_VERSION}-PR${ghprbPullId}-${SPEC}-"
-    } catch (e) {
-        SDK_PREFIX = "OpenJ9-JDK${SDK_VERSION}-${SPEC}-"
-    }
-
-    SDK_SUFFIX = ".tar.gz"
-    TEST_PREFIX = "test-"
-    TEST_SUFFIX = ".tar.gz"
-
+
```
Comment (AdamBrousseau) [inline buildenv/jenkins/common/variables-functions]: some whitespaces here

File: buildenv/jenkins/common/pipeline-functions [hunk 8]
```diff
@@ -252,7 +275,13 @@ def workflow(SDK_VERSION, SPEC, SHAS, OPENJDK_REPO, OPENJDK_BRANCH, OPENJ9_REPO,
             def TEST_JOB_NAME = "Test-${name}-JDK${SDK_VERSION}-${SPEC}"
             testjobs["${TEST_JOB_NAME}"] = {
                 // run tests against the SDK build in the upstream job: Build-JDK${SDK_VERSION}-${SPEC}
-                jobs["${TEST_JOB_NAME}"] = build_with_one_upstream(TEST_JOB_NAME, BUILD_JOB_NAME, jobs["build"].getNumber(), params.VARIABLE_FILE, params.VENDOR_REPO, params.VENDOR_BRANCH, params.VENDOR_CREDENTIALS_ID, params.TEST_NODE, OPENJ9_REPO, OPENJ9_BRANCH, SHAS['OPENJ9'], VENDOR_TEST_REPOS, VENDOR_TEST_BRANCHES, VENDOR_TEST_SHAS, VENDOR_TEST_DIRS, USER_CREDENTIALS_ID, BUILD_LIST)
+                     if (ARTIFACTORY_SERVER) {
+                          def CUSTOMIZED_SDK_URL = "${ARTIFACTORY_BASE_URL}/${BUILD_JOB_NAME}/${jobs["build"].getNumber()}/${SDK_FILENAME}"
+                          echo "Using CUSTOMIZED_SDK_URL = ${CUSTOMIZED_SDK_URL}, ARTIFACTORY_CREDS = ${ARTIFACTORY_CREDS}"
```
Comment (AdamBrousseau) [inline buildenv/jenkins/common/pipeline-functions]: extra whitespace at the end of the line

File: buildenv/jenkins/common/pipeline-functions [hunk 9]
```diff
@@ -252,7 +275,13 @@ def workflow(SDK_VERSION, SPEC, SHAS, OPENJDK_REPO, OPENJDK_BRANCH, OPENJ9_REPO,
...
-                jobs["${TEST_JOB_NAME}"] = build_with_one_upstream(TEST_JOB_NAME, BUILD_JOB_NAME, jobs["build"].getNumber(), params.VARIABLE_FILE, params.VENDOR_REPO, params.VENDOR_BRANCH, params.VENDOR_CREDENTIALS_ID, params.TEST_NODE, OPENJ9_REPO, OPENJ9_BRANCH, SHAS['OPENJ9'], VENDOR_TEST_REPOS, VENDOR_TEST_BRANCHES, VENDOR_TEST_SHAS, VENDOR_TEST_DIRS, USER_CREDENTIALS_ID, BUILD_LIST)
+                     if (ARTIFACTORY_SERVER) {
+                          def CUSTOMIZED_SDK_URL = "${ARTIFACTORY_BASE_URL}/${BUILD_JOB_NAME}/${jobs["build"].getNumber()}/${SDK_FILENAME}"
+                          echo "Using CUSTOMIZED_SDK_URL = ${CUSTOMIZED_SDK_URL}, ARTIFACTORY_CREDS = ${ARTIFACTORY_CREDS}" 
+                          jobs["${TEST_JOB_NAME}"] = build_with_artifactory(TEST_JOB_NAME, params.VARIABLE_FILE, params.VENDOR_REPO, params.VENDOR_BRANCH, params.VENDOR_CREDENTIALS_ID, params.TEST_NODE, OPENJ9_REPO, OPENJ9_BRANCH, SHAS['OPENJ9'], VENDOR_TEST_REPOS, VENDOR_TEST_BRANCHES, VENDOR_TEST_SHAS, VENDOR_TEST_DIRS, USER_CREDENTIALS_ID, BUILD_LIST, CUSTOMIZED_SDK_URL, ARTIFACTORY_CREDS)
+                     } else {
```
Comment (AdamBrousseau) [inline buildenv/jenkins/common/pipeline-functions]: extra whitespace at the end of the line

File: buildenv/jenkins/common/build [hunk 10]
```diff
@@ -261,10 +261,28 @@ def build() {
...
+                server.upload spec: uploadSpec, buildInfo: buildInfo
+                server.publishBuildInfo buildInfo
```
Comment (AdamBrousseau) [inline buildenv/jenkins/common/build]: extra whitespace at the end of the line

File: buildenv/jenkins/common/variables-functions [hunk 11]
```diff
@@ -378,6 +368,17 @@ def set_build_variables() {
...
 
+def set_sdk_variables() {
+    SDK_FILENAME = params.SDK_FILENAME
+    if (!SDK_FILENAME) {
+        DATESTAMP = sh (
+            script: 'date +%Y%d%m%H%M',
+            returnStdout: true
+            ).trim()
+        SDK_FILENAME = "OpenJ9-JDK${SDK_VERSION}-${SPEC}-${DATESTAMP}.tar.gz"
+    }
```
Comment (AdamBrousseau) [inline buildenv/jenkins/common/variables-functions]: Wouldn't hurt to add an echo of the filename here outside the IF

File: buildenv/jenkins/common/build [hunk 12]
```diff
@@ -261,10 +261,28 @@ def build() {
...
-            archiveArtifacts artifacts: "**/${SDK_PREFIX}*${SDK_SUFFIX}", fingerprint: true, onlyIfSuccessful: true
+              sh "tar -C build/${RELEASE}/images/ -zcvf ${SDK_FILENAME} ${JDK_FOLDER}"
+              if (ARTIFACTORY_SERVER) {
+                def server = Artifactory.server ARTIFACTORY_SERVER
+                // if the archive location is modified the definition of CUSTOMIZED_SDK_URL will also need to be updated
+                def uploadSpec = """{
+                  "files":[
+                    {
+                      "pattern": "${env.WORKSPACE}/${SDK_FILENAME}",
+                      "target": "${ARTIFACTORY_REPO}/${env.JOB_NAME}/${env.BUILD_ID}/"
```
[the remark this answers is not in the document]
Comment (AdamBrousseau) [reply buildenv/jenkins/common/build]: I think Joe decided to hardcode it in the Variable file. By default, JENKINS_URL is a full https url so you need to parse it in order get "JENKINS_SERVER". The latter would be more future proof. I believe I've seen your code for this somewhere.

[the remark this answers is not in the document]
Comment (llxia) [reply buildenv/jenkins/common/build]: yes, JENKINS_URL is a full https url which is not ideal to use as folder name. Below is the code that I used before.

```
        def strs = env.HUDSON_URL.split( ':' )
        def url = strs[1].split( '//' )
        def serverName = "undefined"
        if (url && url.length>1) {
            serverName = url[1]
        }
```

[the remark this answers is not in the document]
Comment (jdekonin) [reply buildenv/jenkins/common/build]: I think that is enforcing a directory structure through code that might not be desired.  Setting that via the variable file promotes flexibility of use.  Example...

```
#========================================#
# Artifactory settings
#========================================#
artifactory_server: 'artifactoryserver.companyname.com
artifactory_repo: '<repo name>[/<jenkins server or other directory naming if desired>]'
artifactory_creds: 'acbd1234-a1b2-1234-5678-123abc456def'
```

**Comment by jdekonin** [comment]
Last push 29f1050 is a rebase of master plus the minor whitespace and echo output @AdamBrousseau pointed out.

File: buildenv/jenkins/common/build (hunk 12 above)
[the remark this answers is not in the document]
Comment (AdamBrousseau) [reply buildenv/jenkins/common/build]: What if we had a combination of the 2?
i.e. Write a function to populate JENKINS_SERVER as a global variable. Then use the variable name in the yml file.

[the remark this answers is not in the document]
Comment (llxia) [reply buildenv/jenkins/common/build]: Yes, I am ok with it. All I am saying is that by including Jenkins server name, we will be more future proof. I have already encountered similar issue in Test Result Summary Service, where we need to monitor two Jenkins servers and they have same build names. :)

**Pull request merged**